lib/collection/src/shards/queue_proxy_shard.rs

```diff
@@ -53,9 +51,13 @@ pub struct QueueProxyShard {
 }
 
 impl QueueProxyShard {
-    pub async fn new(wrapped_shard: LocalShard, remote_shard: RemoteShard) -> Self {
+    pub fn new(
+        wrapped_shard: LocalShard,
+        remote_shard: RemoteShard,
+        max_ack_version: Arc<AtomicU64>,
```

Comment:
I don't like that we expose the `max_ack_version` here. I get that it's to make the function non-async.

But I don't think we should expose this to the outside.

Can we find some middle ground here?

Reply:
I don't like it either, but I don't see other way to do it.

And it's either make `QueueProxyShard::now` sync, or `ShardReplicaSet::queue_proxify_local` is *not* cancel safe and there are cancellation tokens everywhere. :/

Reply:
Yeah this is tricky.

Does using explicit cancellation tokens in `queue_proxify_local` look that bad though? Because to be honest, I don't like all the restructuring we're required to do just for this any better.

Reply:
Hmmm... Maybe you're right. We'd have to propagate cancellation token:

- from the shard transfer task to `transfer_shard`
- from `transfer_shard` to `transfer_snapshot`
- and refactor `transfer_snapshot` so that it would `cancel_on_token` everything before and after `queue_proxify_local` and propagate into it

Which might be reasonable.

Reply:
I guess we should keep it as-is for now, since changing this again may be a lot of work. We can revisit this idea later. The current implementation seems logically correct to me. :+1: